Store Parameter Values Set by User on a per-user basis

adam-james-v opened the pull request:
Related: #28181 
and the notion doc:
https://www.notion.so/metabase/Tech-Maintain-user-level-state-in-dashboards-for-filters-fc16909a3216482f896934dd94b54f9a

Adds `:last_used_param_values` key to the `GET /api/dashboard/:id` endpoint.


Still to do:

- [ ] validate the table/model design (getting reviews from backend)
- [x] hook up the endpoints mentioned in the doc (2 in api/dashboard)
- [x] return the user specific parameter values on /api/dashboard/:dashboardID endpoints
- [x] write a few tests to capture the intent of this feature

adam-james-v committed:
Store Parameter Values Set by User on a per-user basis

This is a WIP for #28181 and the notion doc:

https://www.notion.so/metabase/Tech-Maintain-user-level-state-in-dashboards-for-filters-fc16909a3216482f896934dd94b54f9a

Still to do:

- [ ] validate the table/model design
- [ ] hook up the endpoints mentioned in the doc (2 in api/dashboard)
- [ ] return the user specific parameter values on /api/dashboard/:dashboardID endpoints
- [ ] write a few tests to capture the intent of this feature

adam-james-v committed:
Merge branch 'master' into save-dashboard-parameter-values-per-user

adam-james-v committed:
Accidentally deleted a digit in the change ID timestamp

adam-james-v committed:
first pass at writing user param values to the db.

It's in a doseq here which is probably not the correct way to go yet, but it's a step in the right direction.

adam-james-v committed:
Merge branch 'master' into save-dashboard-parameter-values-per-user

adam-james-v committed:
Hydrate dashboard response with `:last_used_param_values` key

If the user has previously set a parameter's value, it will show up in the map under that parameter id.

If the user has no parameter value set, that entry will be 'null'.

adam-james-v committed:
Use proper fn name

adam-james-v committed:
Merge branch 'master' into save-dashboard-parameter-values-per-user

adam-james-v committed:
Only save or retreive user params if there's a current user id

adam-james-v committed:
Merge branch 'save-dashboard-parameter-values-per-user' of https://github.com/metabase/metabase into save-dashboard-parameter-values-per-user

adam-james-v committed:
Add model to necessary lists

adam-james-v committed:
Only run query when there are parameter ids to run it with

adam-james-v committed:
Add a test to confirm that last_used_param_values works

escherize commented on src/metabase/models/user_parameter_value.clj:
```diff
@@ -0,0 +1,51 @@
+  "Upsert or delete parameter value set by the user."
+  [user-id      :- ms/PositiveInt
+   parameter-id :- ms/NonBlankString
+   value        #_#_:- [:maybe :Any]]
```
:Any isn't a malli schema and `[:maybe :any]` is redundant, since nil is in :any. you can just leave off the `:- ...` bit:
```suggestion
   value]
```

escherize replied:
Is it worth actually deleting the user parameter value record when `value` is nil? The docstring mentions this can delete the parameter, but [looking here](https://github.com/camsaul/toucan2/blob/master/test/toucan2/update_test.clj) I don't think the call to t2/update! can. Doesn't that just set the value to nil?

escherize commented on resources/migrations/001_update_migrations.yaml:
```diff
@@ -6137,6 +6137,44 @@ databaseChangeLog:
+                  constraints:
+                    nullable: false
+              - column:
+                  name: value
```
If this is just a string, I don't think you should be using mi/transform-json on the value field.

escherize commented on test/metabase/api/dashboard_test.clj:
```diff
@@ -399,6 +399,39 @@
                             :dashboard_id       dashboard-id}]
           (is (#'api.dashboard/get-dashboard dashboard-id)))))))
 
+(deftest last-used-parameter-value-test
```
Could you add a test for deleting one of these, too?

escherize reviewed (changes requested):
Looks good, just a few things I'd like to see you address

adam-james-v committed:
Add models test namespace for CRUD test

adam-james-v committed:
Merge branch 'master' into save-dashboard-parameter-values-per-user

adam-james-v committed:
The hydration is checked in the dashboard api test already

adam-james-v committed:
Merge branch 'master' into save-dashboard-parameter-values-per-user

adam-james-v committed:
Add idx to FK for user parameter values table + fix tests

PG needs idx to be manually added to foreign keys, so I added that migration.

Additionally, I think I fixed the tests failing by using a temp empty db. Hopefully that fixes the failures.

adam-james-v committed:
maybe this'll be the proper way to write the tests.

Man, I struggled. Oops.

adam-james-v committed:
user mt/id for the user everywhere... silly mistake to miss

adam-james-v committed:
Merge branch 'master' into save-dashboard-parameter-values-per-user
